### ISSUE TITLE
Normalizer looping

### ISSUE THREAD

**Issue by mtzguido** [FStarLang/FStar#1103]
This definition causes the typechecker to loop

```
let f =
    let x = let rec __f (x : int) : int = __f x in
    __f
    in x
```

My guess is that `__f` is not annotated as effectful (yet) and the normalizer just applies it as any other function. Annotating `__f` as `Dv` stops that, and the file verifies. Taking the seemingly needless `let x =.. in x` out causes the file to fail typechecking due to non-termination, as expected.

Looking into it now...

**Comment by mtzguido** [comment]
Indeed, this happens at generalization time, where the function is assumed total. Totality will be checked by a _later_ SMT query, and fail. (If one changes the rightmost body from `x` to, say, `8` which doesn't depend on `__f`,  generalization works and we then get the expected totality failure).

Yet the normalizer assumes that any total function is total, which would be the case if the let rec was top-level, since it would have already been checked..

**Comment by mtzguido** [comment]
Simple fix that doesn't seem to break anything: exclude fixpoint reductions during generalization.

Heads up @nikswamy, in case this is not desirable, but closing for now.

**Comment by mtzguido** [comment]
Aaaand that's not enough. Even if generalization won't loop now, we can still loop on typechecking, as the following example shows.

```
let g (u:unit) = ()

let f =
    let x =
        let rec __f (x : int) : Type = __f x in
        __f
    in (() <: (x 0)) <: int
```

The annotations are just a way of mentioning the recursive function in a type, which of course are normalized heavily. Disabling fixpoint reductions here would be too restrictive, and break things SimplePrintf.

I'm not sure on what to do here. A possibility is to eagerly check for totality of `let rec`s, but that goes against what we discussed on Slack just yesterday about postponing queries until all uvars are instantiated. Or at least, means we need to force resolution of all uvars for let recs. Another is to distinguish previously-checked let recs from the ones in the current top-level, and reduce only the former.

(Running into this when constructing `let rec`s from tactics, mostly)

**Comment by mtzguido** [comment]
Spoke to Nik about this a while ago, and he correctly pointed out that neither of those help when in an inconsistent environment. i.e., this loops
```
let rec r (f:False) (x:int) : Tot int =
    r f x

let l (f:False) =
    assert_norm (r f 1 == 1)
```
We thought of adding some hard (possibly configurable) limit to Zeta reductions, as an heuristic stop which would abort reduction and then catch all of these cases. For the previous example, we could still implement any of the other solutions (so we never take a single step), but a catch-all loop-break seems desirable.

**Comment by nikswamy** [comment]
the issue has flip-flopped a bit, so not sure what's the current status. @mtzguido  can you summarize?

**Comment by mtzguido** [comment]
(That last close and reopen was me missing the button, I think)

Status is the same, I think. The second example on this file loops, but I really think it should not take a single step. I didn't get around to implementing the Zeta-limit either.

Also a few days ago I came up with this example, which seems bad too as it loops in consistent environments (the last example shouldn't even typecheck?).
```
module T

let rec g (_ : squash False) : Tot unit =
    g ()

(* Stack overflow *)
let _ = assert_norm (g == g)

(* Infinite loop *)
let _ = assert_norm (g () == g ())
```

**Comment by mtzguido** [comment]
The second example no longer loops. I don't think there's an easy way to fix the `assert_norm` examples, since `assert_norm` is meant to explicitly normalize it argument *in order to* compute a cleaner VC, hence it happens before the function is verified, and we cannot detect that `g ()` doesn't verify in the second one. Even if we could detect those situations, it wouldn't fix the first example `g == g`. I suppose `assert_norm` is a power tool, hence the user should be wary of this.

Closing, feel free to disagree and reopen.
